examples/cpu-hello-world/main.cpp

```diff
@@ -0,0 +1,210 @@
+// main.cpp
+
+#include <stdio.h>
+
+// This file implements an extremely simple example of loading and
+// executing a Slang shader program on the CPU. 
+//
+// More information about generation C++ or CPU code can be found in docs/cpu-target.md
+//
+// NOTE! This test will only run on a system correctly where slang can find a suitable
+// C++ compiler - such as clang/gcc/visual studio
+//
+// The comments in the file will attempt to explain concepts as
+// they are introduced.
+//
+// Of course, in order to use the Slang API, we need to include
+// its header. We have set up the build options for this project
+// so that it is as simple as:
+#include <slang.h>
+
+// Allows use of ComPtr - which we can use to scope any 'com-like' pointers easily
+#include <slang-com-ptr.h>
+// Provides macros for handling SlangResult values easily 
+#include <slang-com-helper.h>
+
+// This includes a useful small function for setting up the prelude (described more further below).
+#include "../../source/core/slang-test-tool-util.h"
+
+// Slang namespace is used for elements support code (like core) which we use here
+// for ComPtr<> and TestToolUtil
+using namespace Slang;
+
+// Slang source is converted into C++ code which is compiled by a backend compiler.
+// That process uses a 'prelude' which defines types and functions that are needed
+// for everything else to work.
+// 
+// We include the prelude here, so we can directly use the types as were used by the
+// compiled code. It is not necessary to include the prelude, as long as memory is
+// laid out in the manner that the generated slang code expects. 
+#define SLANG_PRELUDE_NAMESPACE CPPPrelude
+#include "../../prelude/slang-cpp-types.h"
```

Comment:
That is an interesting point. I agree a client should be able roll their own, and they can - we spec the layout. It is perhaps arguably 'odd' for client code to potentially include 'prelude code'. But in practice, both in this example and in other test code - somewhere there should be a definition of these types (if not RwStructuredBuffer - for ITexture2D etc). We could have another set of headers that client code includes - but then we are repeating ourselves. We could have nothing at all. But then our example code will have to define something somewhere - so we'd repeat ourselves.
The prelude it split - such that client code does not have to include all of the prelude. 
It can be documented in code and elsewhere more that rolling your own is fine. But in practice this seems to be a reasonable approach such that it is.

Reply:
I guess the option that I'd favor most is:

* There is a `slang-cpu-target-extras.h` or similar header (with a better name...) that is intended for client code to include when interacting with code generated by our CPU back-end. Exactly what gets defined in that file is decided by how "deep" we want the interop to be (in the limit we start exposing `slang::float3`, etc., with all its operators, but that starts to feel like an over-reach).

* Our C++ prelude code might include/embed `slang-cpu-target-extras.h` as a convenient way to get definitions of all the ABI-impacting types that should guarantee a compatible layout. Even so, the fact that our CPU codegen happens to use the same header that client code is pushed to use is an implementation detail.

If we can make that option work then we both avoid the code duplication *and* keep the user from having to think about the prelude until/unless they want to roll a custom one (which is on my mental list of degrees of freedom I don't see 99% of users taking advantage of).